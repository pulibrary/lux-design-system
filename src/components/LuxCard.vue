<template>
  <article
    :id="id"
    @click.capture="select($event)"
    class="lux-card"
    :class="[
      size,
      { 'lux-card-selected': selected },
      { 'lux-card-edited': edited },
      { 'lux-card-disabled': disabled },
    ]"
    v-bind:style="{ 'max-width': cardPixelWidth + 'px' }"
  >
    <slot></slot>
  </article>
</template>

<script>
/**
 * Cards are used to apply a container around a related grouping of information.
 */
export default {
  name: "LuxCard",
  status: "ready",
  release: "1.0.0",
  type: "Element",
  props: {
    /**
     * Sets the id to reference this card with.
     */
    id: {
      type: String,
      default: "",
    },
    /**
     * Sets the URL linking to the card content
     */
    cardUrl: {
      type: String,
      default: "",
    },
    /**
     * Sets arbitrary card width. It's recommended to use size over this setting.
     */
    cardPixelWidth: {
      default: "",
    },
    /**
     * Sets the size of the card `small, medium, large, full-width`
     */
    size: {
      type: String,
      default: "medium",
      // validator: (value) => {
      //   return value.match(/(small|medium|large|full-width)/);
      // },
    },
    /**
     * Indicates whether the card is selected.
     */
    selected: {
      type: Boolean,
      default: false,
    },
    /**
     * Indicates whether the object a card represents has been altered from its persisted form.
     */
    edited: {
      type: Boolean,
      default: false,
    },
    /**
     * Indicates the user cannot interact with the card.
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * Indicates the user cannot interact with the card.
     */
    cardTitle: {
      type: String,
      default: "",
    },
  },
  methods: {
    select: function (event) {
      this.$emit("card-click", event)
    },
  },
}
</script>

<style lang="scss" scoped>
@import "../assets/styles/variables.css";
@import "../assets/styles/mixins.scss";
@import "../assets/styles/spacing.scss";

.lux-card {
  font-family: var(--font-family-text);
  line-height: var(--line-height-base);
  background: var(--color-white);
  box-shadow: var(--box-shadow-small);
  color: var(--color-rich-black);
  position: relative;

  &:active {
    cursor: move;
  }
  &:focus,
  &-selected {
    box-shadow: var(--box-shadow-selected);
  }
  &-edited {
    background: var(--color-princeton-orange-on-white);
  }

  &-disabled {
    opacity: 0.2;
    cursor: not-allowed;
  }
  &.small {
    width: 200px;
  }
  &.medium {
    width: 300px;
  }
  &.large {
    width: 400px;
  }

  .lux-heading,
  .lux-text-style {
    padding: 0 1rem;
  }
  .lux-heading,
  .lux-text-style,
  .lux-media-image {
    pointer-events: none;
  }
  .lux-text-style {
    padding-bottom: 1rem;
  }
  :deep(.lux-tag-item) {
    border-radius: var(--border-radius-pill);
    font-weight: 400;
  }
  .full-width {
    width: 100%;
    display: flex;
    align-items: flex-start;
    flex-flow: row wrap;
    .lux-card-media + .lux-card-header {
      padding-left: 0;
    }
    .lux-card-media + .lux-card-content,
    .lux-card-content:only-child {
      flex: 1;
    }
    :deep(.lux-card-content) {
      padding: 1rem;
      @media (min-width: 900px) {
        padding: var(--space-base);
      }
    }
    @media (min-width: 600px) {
      .lux-card-header {
        flex: 1;
      }
    }
  }
  .lux-card :deep(.lux-link) {
    color: var(--color-rich-black);
    outline: 0;
    &::after {
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      z-index: 1;
      pointer-events: auto;
      content: "";
      background-color: rgba(0, 0, 0, 0);
    }
    &:hover,
    &:focus {
      box-shadow: none;
      &::after {
        box-shadow: var(--box-shadow-selected);
        transition: box-shadow 0.2s ease;
      }
    }
    &:visited {
      color: var(--color-rich-black);
    }
  }
}
</style>

<docs>
  ```
    <div>
      <!-- Card without sections -->
      <lux-card id="a">
        <media-image src="https://picsum.photos/600/300/?random" height="medium" cover></media-image>
        <lux-heading level="h2">Title</lux-heading>
        <text-style variation="default">Design isn't just about the look and feel. Design is how it works.</text-style>
      </lux-card>

      <!-- Card with sections -->
      <lux-card id="b" size="full-width">
        <card-media>
          <lux-icon-base width="50" height="50" icon-hide="true">
            <lux-icon-globe></lux-icon-globe>
          </lux-icon-base>
        </card-media>
        <card-header>
          <!-- Hyperlink is the text screen readers would read, we don't want to wrap the entire card -->
          <lux-heading level="h2" size="h3"><hyperlink href="#">Code4Lib - Trip ID 1234</hyperlink></lux-heading>
          <text-style>Jan 9, 2019 to Jan 16, 2019</text-style>
        </card-header>
        <card-content>
          <tag type="tag" :tag-items="[
            {name: 'Pending', color: 'yellow', style: 'pill'}
            ]"
            horizontal="end"
            size="small"/>
          <text-style type="span" variation="small">Last Updated on Dec 15, 2018</text-style>
        </card-content>
      </lux-card>
    </div>
  ```
</docs>
