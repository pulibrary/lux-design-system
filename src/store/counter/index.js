export const counterState = {
  count: 10,
}

export const counterMutations = {
  increment(state) {
    // `state` is the local module state
    state.count++
  },
}

export const counterGetters = {
  count: state => {
    return state.count
  },
}
