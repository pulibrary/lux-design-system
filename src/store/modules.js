import { galleryState, galleryMutations } from "./gallery/index"
import { counterState, counterMutations, counterGetters } from "./counter/index"

export const galleryModule = {
  state: galleryState,
  mutations: galleryMutations,
}

export const counterModule = {
  state: counterState,
  mutations: counterMutations,
  getters: counterGetters,
}

let modules
export default modules = { counterModule, galleryModule }
