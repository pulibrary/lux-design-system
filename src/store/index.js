import { createStore } from "vuex"
// import { counterModule, galleryModule } from "./modules"
import { galleryModule } from "./modules"

export default createStore({
  state: { count: 10 },

  mutations: {
    increment(state) {
      // `state` is the local module state
      state.count++
    },
  },

  getters: {
    count: state => {
      return state.count
    },
  },
  modules: {
    // counter: counterModule,
    gallery: galleryModule,
  },
})
