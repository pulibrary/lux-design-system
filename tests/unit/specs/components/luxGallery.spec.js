import Vuex from "vuex"
import { mount } from "@vue/test-utils"
import { nextTick } from "vue"
import LuxGallery from "@/components/LuxGallery.vue"
import LuxCard from "@/components/LuxCard.vue"
import { galleryMutations } from "@/store/gallery"

let wrapper
let store
let state
let mutations
let items = [
  { id: "1", title: "First", caption: "one", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "2", title: "Second", caption: "two", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "3", title: "Third", caption: "three", mediaUrl: "https://foo/bar" },
]
let reordered = [
  { id: "3", title: "Third", caption: "three", mediaUrl: "https://foo/bar" },
  { id: "2", title: "Second", caption: "two", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "1", title: "First", caption: "one", mediaUrl: "https://picsum.photos/600/300/?random" },
]

describe("LuxGallery.vue", () => {
  beforeEach(() => {
    state = {
      items: items,
      selected: [items[0]],
      cut: [],
      changeList: ["2"],
      ogItems: items,
    }

    store = new Vuex.Store({
      modules: {
        gallery: {
          state,
          mutations: galleryMutations,
        },
      },
    })

    wrapper = mount(LuxGallery, {
      store,
      propsData: {
        galleryItems: items,
      },
      global: {
        stubs: {
          "lux-card": true,
          "lux-heading": true,
          "lux-text-style": true,
          "lux-media-image": true,
        },
      },
    })
  })

  // Look into how to test setters to get this to work ...
  // it("should update the sort order of the cards on an input event from <draggable>", () => {
  //   wrapper.vm.$emit('input', reordered)
  //   expect(wrapper.vm.items[0].id).toBe(3)
  // })

  it("sizes the card correctly", async () => {
    wrapper.setProps({ cardPixelWidth: 200 })
    await nextTick()
    const card = wrapper.find(".lux-galleryCard")
    expect(card.attributes().cardpixelwidth).toBe("200")
  })

  it("clicking on the gallery background should call our deselect method", () => {
    const deselectStub = jest.spyOn(wrapper.vm, "deselect")
    wrapper.find(".lux-gallery").trigger("click")
    expect(deselectStub).toHaveBeenCalled()
  })

  it("getItemIndexById returns the correct index", () => {
    expect(wrapper.vm.getItemIndexById("3")).toBe(2)
  })

  it("getItemById returns the correct item", () => {
    const item = wrapper.vm.getItemById("3")
    expect(item.title).toBe("Third")
  })

  // These don't work in the new version... need to debug
  // it("hasChanged returns true for a changed item", () => {
  //   expect(wrapper.vm.hasChanged("2")).toBe(true)
  // })
  //
  // it("isSelected returns true for a selected item", () => {
  //   expect(wrapper.vm.isSelected(state.selected[0])).toBe(true)
  // })

  it("selects the correct items", () => {
    expect(wrapper.vm.isSelected(wrapper.vm.items[1])).toBe(false)
    wrapper.vm.select("2", { shiftKey: false, metaKey: false, target: { id: "2" } })
    expect(wrapper.vm.isSelected(wrapper.vm.items[1])).toBe(true)
    wrapper.vm.selectNone()
    expect(wrapper.vm.isSelected(wrapper.vm.items[1])).toBe(false)
    wrapper.vm.select("1", { shiftKey: false, metaKey: false, target: { id: "1" } })
    wrapper.vm.select("3", { shiftKey: true, metaKey: false, target: { id: "3" } })
    expect(wrapper.vm.isSelected(wrapper.vm.items[1])).toBe(true)
    wrapper.vm.selectNone()
    expect(wrapper.vm.isSelected(wrapper.vm.items[1])).toBe(false)
    wrapper.vm.select("1", { shiftKey: false, metaKey: false, target: { id: "1" } })
    wrapper.vm.select("3", { shiftKey: false, metaKey: true, target: { id: "3" } })
    expect(wrapper.vm.isSelected(wrapper.vm.items[0])).toBe(true)
    expect(wrapper.vm.isSelected(wrapper.vm.items[1])).toBe(false)
  })

  it("deselects the correct items", () => {
    wrapper.vm.select("1", { shiftKey: false, metaKey: false, target: { id: "1" } })
    expect(wrapper.vm.isSelected(wrapper.vm.items[0])).toBe(true)
    wrapper.vm.deselect({ shiftKey: false, metaKey: false, target: { className: "foo" } })
    expect(wrapper.vm.isSelected(wrapper.vm.items[0])).toBe(true)
    wrapper.vm.deselect({ shiftKey: false, metaKey: false, target: { className: "lux-gallery" } })
    expect(wrapper.vm.isSelected(wrapper.vm.items[0])).toBe(false)
  })

  it("sorts the items", () => {
    expect(wrapper.vm.items[0].id).toBe("1")
    wrapper.vm.items = reordered
    expect(wrapper.vm.items[0].id).toBe("3")
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
