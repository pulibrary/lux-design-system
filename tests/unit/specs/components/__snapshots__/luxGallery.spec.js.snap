// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuxGallery.vue has the expected html structure 1`] = `
<div
  class="lux-gallery"
>
  <lux-card
    cardpixelwidth="300"
    class="lux-galleryCard"
    data-draggable="true"
    disabled="false"
    edited="false"
    id="1"
    selected="false"
    size="medium"
  >
    <lux-media-image
      src="https://picsum.photos/600/300/?random"
    />
    <lux-heading
      level="h2"
    >
      First
    </lux-heading>
    <lux-text-style
      variation="default"
    >
      one
    </lux-text-style>
  </lux-card>
  <lux-card
    cardpixelwidth="300"
    class="lux-galleryCard"
    data-draggable="true"
    disabled="false"
    edited="false"
    id="2"
    selected="false"
    size="medium"
  >
    <lux-media-image
      src="https://picsum.photos/600/300/?random"
    />
    <lux-heading
      level="h2"
    >
      Second
    </lux-heading>
    <lux-text-style
      variation="default"
    >
      two
    </lux-text-style>
  </lux-card>
  <lux-card
    cardpixelwidth="300"
    class="lux-galleryCard"
    data-draggable="true"
    disabled="false"
    edited="false"
    id="3"
    selected="false"
    size="medium"
  >
    <lux-media-image
      src="https://foo/bar"
    />
    <lux-heading
      level="h2"
    >
      Third
    </lux-heading>
    <lux-text-style
      variation="default"
    >
      three
    </lux-text-style>
  </lux-card>
</div>
`;
